test/testother.cpp

```diff
@@ -3904,6 +3907,127 @@ class TestOther : public TestFixture {
+        ASSERT_EQUALS("[test.cpp:1] -> [test.cpp:1]: (style) Same expression on both sides of '!='.\n", errout.str());
+
+        check("int a = 1; void f() { if ( a != 1){}}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() { static const int a = 1; if ( a != 1){}}\n");
+        ASSERT_EQUALS("[test.cpp:1] -> [test.cpp:1]: (style) Same expression on both sides of '!='.\n", errout.str());
+
+        check("void f() { static int a = 1; if ( a != 1){}}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() { int a = 1; if ( a != 1){ a++; }}\n");
+        ASSERT_EQUALS("[test.cpp:1] -> [test.cpp:1]: (style) Same expression on both sides of '!='.\n", errout.str());
+
+        check("void f(int b) { int a = 1; while (b) { if ( a != 1){} a++; } } \n");
+        ASSERT_EQUALS("", errout.str());
+    }
+
+    void duplicateExpression8() {
+        check("void f() { int a = 1; int b = a; a = 2; if ( b != a){}}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f(int * a, int i) { int b = a[i]; a[i] = 2; if ( b != a[i]){}}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f(int * a, int i) { int b = *a; *a = 2; if ( b != *a){}}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("struct A { int f() const; };\n"
+              "A g();\n"
+              "void foo() {\n"
+              "    for (const A x = A();;) {\n"
+              "        const int a = x.f();\n"
+              "        x = g();\n"
+              "        if (x.f() == a) break;\n"
+              "    }\n"
+              "}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("int f(int i);\n"
+              "struct A {\n"
+              "    enum E { B, C };\n"
+              "    bool f(E);\n"
+              "};\n"
+              "void foo() {\n"
+              "    A a;\n"
+              "    const bool x = a.f(A::B);\n"
+              "    const bool y = a.f(A::C);\n"
+              "    if(!x && !y) return;\n"
+              "}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void foo() { \n"
+              "    const bool x = a.f(A::B);\n"
+              "    const bool y = a.f(A::C);\n"
+              "    if (!x && !y) return;\n"
+              "}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f(bool * const b);\n"
+              "void foo() { \n"
+              "    bool x = true;\n"
+              "    bool y = true;\n"
+              "    f(&x);\n"
+              "    if (!x && !y) return;\n"
+              "}\n");
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() { const int a = {}; if(a == 1) {} }\n");
+        ASSERT_EQUALS("", errout.str());
+    }
+
+    void duplicateExpressionLoop() {
+        check("void f() { int a = 1; while ( a != 1){}}\n");
+        ASSERT_EQUALS("[test.cpp:1] -> [test.cpp:1]: (style) Same expression on both sides of '!='.\n", errout.str());
```

Comment:
hmm.. I do not define the expressions as "same".

The `CheckCondition::alwaysTrueFalse` writes this warning:

    [1.c:1]: (style) Condition 'a!=1' is always false

and that warning seems more correct as far as I see.

Reply:
But that may use ValueFlow instead of `isSameExpression` to check for this.

Reply:
we try to avoid that duplicate warnings are written. If you can write a better warning message than the "Condition is always false" then I prefer that we keep it and add some exception in the alwaysTrueFalse checker.